cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_nodegroup_test.go

```diff
@@ -699,3 +700,113 @@ func TestNodeGroupMachineSetDeleteNodes(t *testing.T) {
 		t.Fatalf("expected 5 nodes, got %v", actual)
 	}
 }
+
+func TestNodeGroupMachineSetDeleteNodesWithMismatchedNodes(t *testing.T) {
+	makeMachineSet := func(i int, replicaCount int) *v1beta1.MachineSet {
+		name := fmt.Sprintf("machineset-%d", i)
+		return &v1beta1.MachineSet{
+			TypeMeta: v1.TypeMeta{
+				Kind: "MachineSet",
+			},
+			ObjectMeta: v1.ObjectMeta{
+				Name:      name,
+				Namespace: "test-namespace",
+				UID:       types.UID(name),
+			},
+			Spec: v1beta1.MachineSetSpec{
+				Replicas: int32ptr(int32(replicaCount)),
+			},
+		}
+	}
+
+	// Creates a machine and node; links the machine and node.
+	// Additionally links the machine with machineSet.
+	makeLinkedMachine := func(i int, machineSet *v1beta1.MachineSet) (*apiv1.Node, *v1beta1.Machine) {
+		node := &apiv1.Node{
+			TypeMeta: v1.TypeMeta{
+				Kind: "Node",
+			},
+			ObjectMeta: v1.ObjectMeta{
+				Name: fmt.Sprintf("node-%d", i),
+				Annotations: map[string]string{
+					machineAnnotationKey: fmt.Sprintf("test-namespace/machine-%d", i),
+				},
+			},
+			Spec: apiv1.NodeSpec{
+				ProviderID: fmt.Sprintf("providerid-%d", i),
+			},
+		}
+
+		machine := &v1beta1.Machine{
+			TypeMeta: v1.TypeMeta{
+				Kind: "Machine",
+			},
+			ObjectMeta: v1.ObjectMeta{
+				Name:      fmt.Sprintf("machine-%d", i),
+				Namespace: "test-namespace",
+				OwnerReferences: []v1.OwnerReference{{
+					Name: machineSet.Name,
+					Kind: machineSet.Kind,
+					UID:  machineSet.UID,
+				}},
+			},
+			Status: v1beta1.MachineStatus{
+				NodeRef: &apiv1.ObjectReference{
+					Kind: node.Kind,
+					Name: node.Name,
+				},
+			},
+		}
+
+		return node, machine
+	}
+
+	nreplicas := 1
+	machineSet0 := makeMachineSet(0, nreplicas)
+	machineSet1 := makeMachineSet(1, nreplicas)
+	node0, machine0 := makeLinkedMachine(0, machineSet0)
+	node1, machine1 := makeLinkedMachine(1, machineSet1)
+
+	controller, stop := mustCreateTestController(t, testControllerConfig{
+		nodeObjects: []runtime.Object{
+			node0,
+			node1,
+		},
+		machineObjects: append([]runtime.Object{},
+			machine0, machineSet0,
+			machine1, machineSet1),
+	})
+	defer stop()
+
+	ng0, err := newNodegroupFromMachineSet(controller, machineSet0)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	ng1, err := newNodegroupFromMachineSet(controller, machineSet1)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	// Deleting nodes that are not in ng0 should fail.
+	if err := ng0.DeleteNodes([]*apiv1.Node{node1}); err == nil {
+		t.Error("expected an error")
```

Comment:
Would make sense to check for specific error in all `ng0.DeleteNodes` and `ng1.DeleteNodes`.

Reply:
Addressed in 0afd39c7709844631d5281a2ebd4e611316d9502